NIFI-10108 Processor scheduling via parameter

timeabarna opened the pull request:
# Summary

[NIFI-10108](https://issues.apache.org/jira/browse/NIFI-10108) enables Processors' Scheduling Period to be configured via parameters. A use case may be a generic flow (extracting data from a table) should run for table A every hour in environment A but for table B every day in another environment. Using parameters it can be easily achievable without modifying Processor configuration.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI-10108) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

timeabarna commented:
Hello @markap14, Can you please review?

markap14 commented:
Thanks @timeabarna will do.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java:
```diff
@@ -1928,6 +1967,21 @@ public void setMaxBackoffPeriod(String maxBackoffPeriod) {
+        final ParameterTokenList parameterTokenList = parameterParser.parseTokens(schedulingPeriod);
+        final String evaluatedValue = parameterTokenList.substitute(parameterContext);
+
+        if (evaluatedValue != null && !evaluatedValue.startsWith("#{")) {
+            return evaluatedValue;
+        } else {
+            return schedulingStrategy.getDefaultSchedulingPeriod();
+        }
```
After more fully reviewing this PR, I would say this method doesn't really have anything to do with scheduling period. It has to do with evaluating parameters. So perhaps we should just name it `evaluateParameters` or `substituteParameters`.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java:
```diff
@@ -505,24 +510,30 @@ public String getSchedulingPeriod() {
         }
 
+        final String evaluatedSchedulingPeriod = evaluateSchedulingPeriodFromParameter(schedulingPeriod);
+        final ParameterParser parameterParser = new ExpressionLanguageAgnosticParameterParser();
+        final ParameterTokenList parameterTokenList = parameterParser.parseTokens(schedulingPeriod);
+
+        parameterReferences = new ArrayList<>(parameterTokenList.toReferenceList());
+
         switch (schedulingStrategy) {
         case CRON_DRIVEN: {
             try {
-                new CronExpression(schedulingPeriod);
+                new CronExpression(evaluatedSchedulingPeriod);
             } catch (final Exception e) {
                 throw new IllegalArgumentException(
-                        "Scheduling Period is not a valid cron expression: " + schedulingPeriod);
+                        "Scheduling Period is not a valid cron expression: " + evaluatedSchedulingPeriod);
             }
         }
             break;
         case PRIMARY_NODE_ONLY:
         case TIMER_DRIVEN: {
-            final long schedulingNanos = FormatUtils.getTimeDuration(requireNonNull(schedulingPeriod),
+            final long schedulingNanos = FormatUtils.getTimeDuration(requireNonNull(evaluatedSchedulingPeriod),
                     TimeUnit.NANOSECONDS);
             if (schedulingNanos < 0) {
                 throw new IllegalArgumentException("Scheduling Period must be positive");
```
Because these values can change outside of this method now, by changing a parameter, we need to perform these checks during validation. Because we'll be checking it there, and because we shouldn't throw an Exception if referencing a parameter that hasn't yet been set, the validation should now be removed from this method.

markap14 reviewed (changes requested):
Hey @timeabarna thanks for putting this up! I think this is a good start, but there are a few things that need to be adjusted. I made several comments inline. Most of them very minor (like use `Collections.emptyList()` over `Collections.EMPTY_LIST`).

Unless I missed it, I think we also need to be sure to update `StandardParameterReferenceManager` so that its `isComponentReferencing(ComponentNode componentNode, String parameterName)` method also takes this into account.

timeabarna commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java:
```diff
@@ -505,24 +510,30 @@ public String getSchedulingPeriod() {
             } catch (final Exception e) {
                 throw new IllegalArgumentException(
-                        "Scheduling Period is not a valid cron expression: " + schedulingPeriod);
+                        "Scheduling Period is not a valid cron expression: " + evaluatedSchedulingPeriod);
```
Corrected also in StandardProcessorDao.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java:
```diff
@@ -1209,33 +1194,112 @@ protected Collection<ValidationResult> computeValidationErrors(final ValidationC
+            final String schedulingPeriod = getSchedulingPeriod();
+            final String evaluatedSchedulingPeriod = evaluateParameter(schedulingPeriod);
+
+            if (evaluatedSchedulingPeriod != null) {
```
This should never be `null`, should it?

timeabarna replied:
For User experience I would not like to check validation at first error found. As parameter substitution can be null if a parameter can not be substituted, to be able to continue validation checks, I need to check for null here. 
If the value is incorrect the Processor will be invalid and no null value will be set.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/StandardProcessorNode.java:
```diff
@@ -1969,17 +2033,12 @@ public void setMaxBackoffPeriod(String maxBackoffPeriod) {
 
     @Override
-    public String evaluateSchedulingPeriodFromParameter(final String schedulingPeriod) {
+    public String evaluateParameter(final String parameter) {
```
I don't think the argument here is really a parameter is it? It's some value that may or may not contain parameters. Perhaps should call it `evaluateParameters` and call the argument `value`?

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/parameter/StandardParameterReferenceManager.java:
```diff
@@ -128,6 +128,14 @@ private boolean isComponentReferencing(final ComponentNode componentNode, final
             }
         }
 
+        if (componentNode instanceof ProcessorNode) {
```
We shouldn't be treating ProcessorNode specially. Perhaps it makes more sense to get rid of this `isComponentReferencing` method from `StandardParameterReferenceManager` and instead add a `isReferencingParameter(String parameterName)` to `ComponentNode`. Then the existing implementation would move to `AbstractComponentNode` and `StandardProcessorNode` can then override the method, call the `super.isReferencingParameter` and if false check its `configurationParameterReferences`

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/connectable/Connectable.java:
```diff
@@ -311,5 +311,5 @@ default boolean isSessionBatchingSupported() {
     void setMaxBackoffPeriod(String maxBackoffPeriod);
 
-    String evaluateSchedulingPeriodFromParameter(String schedulingPeriod);
+    String evaluateParameter(String parameter);
```
As commented, based on the way this is used, I think the signature should be `String evaluateParameters(String value);`

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -1084,12 +1092,29 @@ public void onParametersModified(final Map<String, ParameterUpdate> updatedParam
         }
     }
 
+    protected void increaseReferenceCounts(final String parameterName) {
+        parameterReferenceCounts.merge(parameterName, 1, (a, b) -> a == -1 ? null : a + b);
+    }
+
+    protected void decreaseReferenceCounts(final String parameterName) {
```
Given that this is always changing the value by `1`, it probably is best to call these `incrementReferenceCounts` and `decrementReferenceCounts`

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -1084,12 +1092,29 @@ public void onParametersModified(final Map<String, ParameterUpdate> updatedParam
+     * Verifies whether any referenced configuration parameter has been updated and its effective value has changed.
+     * If yes, the component needs to be re-validated.
+     * @param updatedParameters updated parameters
+     * @return true if the component requires re-validation
```
With a method named `isConfigurationParameterModified`, it should not return `true` to indicate that the component requires validation. That's simply a side-effect. It should return `true` if any of the given parameters are used in its configuration.

timeabarna replied:
Due to refactor this method has been removed

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/ProcessorNode.java:
```diff
@@ -300,4 +301,12 @@ public abstract CompletableFuture<Void> stop(ProcessScheduler processScheduler,
+     *
+     * @return the configuration Parameter references for this Processor
+     */
+    public abstract List<ParameterReference> getConfigurationParameterReferences();
```
Not sure that the naming convention here makes sense. Parameters are always used for configuration. Processor properties are very much configuration.

timeabarna replied:
This method has been removed.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java:
```diff
@@ -1462,6 +1459,25 @@ private Set<AffectedComponentEntity> getComponentsAffectedByParameterContextUpda
         return dtoFactory.createAffectedComponentEntities(affectedComponents, revisionManager);
     }
 
+    private void getAffectedComponentsByParameterUpdate(final ParameterContextDTO parameterContextDto, final Set<String> updatedParameterNames, final Set<ComponentNode> affectedComponents,
```
We should not have a getter that doesn't return a value. Is the intent to populate the provided `affectedComponents` set? If so, perhaps that should be returned instead.

timeabarna replied:
This method has been removed.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java:
```diff
@@ -1462,6 +1459,25 @@ private Set<AffectedComponentEntity> getComponentsAffectedByParameterContextUpda
         return dtoFactory.createAffectedComponentEntities(affectedComponents, revisionManager);
     }
 
+    private void getAffectedComponentsByParameterUpdate(final ParameterContextDTO parameterContextDto, final Set<String> updatedParameterNames, final Set<ComponentNode> affectedComponents,
```
Perhaps this should be `getComponentsAffectedBy...`

timeabarna replied:
This method has been removed.

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java:
```diff
@@ -1084,12 +1092,29 @@ public void onParametersModified(final Map<String, ParameterUpdate> updatedParam
+     * @param updatedParameters updated parameters
+     * @return true if the component requires re-validation
+     **/
+    public abstract boolean isConfigurationParameterModified(final Map<String, ParameterUpdate> updatedParameters);
```
I'm not sure exactly what a "configuration parameter" is - i presume here we mean a parameter that is used within the configuration, other than properties, such as the Scheduling Period.  But not sure that this comes across from the code. Also not entirely sure that this is something that we should distinguish between - parameters that are used by properties and parameters that are used elsewhere. Perhaps this would call for a slight refactoring elsewhere but not sure that we should distinguish between the two. Should we just have a `Set<String> getReferencedParameterNames()` on ComponentNode? Would that give us everything that we need?

timeabarna replied:
This method has been removed.

markap14 reviewed (changes requested):
Thanks for the update @timeabarna . I think we're getting close to where we need to be. I had a few concerns about the naming conventions that I mentioned inline. I think we can clarify some things to make it read a bit more clearly. I also had a few questions about the abstractions that we have here. Commented more inline.

timeabarna commented:
Hello @markap14, 
Thanks for your recommendations. I've refactored the code so most of the newly created methods have been removed. I've refactored the code in onParametersModified() method and moved the component referncing check from StandardParameterreferencingManager.class to AbstractComponentNode.class.

Can you please review the modifications?
Thanks in advance.

markap14 commented:
Thanks for updating @timeabarna. I think all of my concerns have been addressed. I did some testing and everything appears to be working as expected. I think there's some potential in the future for some improvements to the UI in order to allow auto-completion like we do in properties. But that can be done in a follow-on effort. Things look good here from my perspsective.
+1 will merge to main.
